perf-tool/src/objectalloc.cpp

```diff
@@ -0,0 +1,150 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+// TODO: object count
+#include <jvmti.h>
+#include <string.h>
+
+#include "agentOptions.hpp"
+#include "infra.hpp"
+#include "json.hpp"
+#include "server.hpp"
+#include "infra.hpp"
+
+#include <iostream>
+#include <chrono>
+#include <ctime>
+#include <iostream>
+#include <iomanip>
+#include <ctime>
+#include <chrono>
+#include <atomic>
+
+using json = nlohmann::json;
+using namespace std::chrono;
+
+std::atomic<bool> objAllocBackTraceEnabled {true};
+std::atomic<int> objAllocSampleCount {0};
+std::atomic<int> objAllocSampleRate {1};
+
+// Enables or disables the back trace option if sampleRate == 0
+void setObjAllocBackTrace(bool val){
+    objAllocBackTraceEnabled = val;
+    return;
+}
+
+// to turn off backTrace, set sampleRate to 0
+void setObjAllocSampleRate(int rate) {
+    if (rate > 0) {
+        objAllocBackTraceEnabled = true;
+        objAllocSampleRate = rate;
+    } else {
+        objAllocBackTraceEnabled = false;
+    }
+    return;
+}
+
+/*** retrieves object type name, size (in bytes), allocation rate (bytes/microsec),
+ *      and backtrace for every nth sample (if enabled)                             ***/
+JNIEXPORT void JNICALL VMObjectAlloc(jvmtiEnv *jvmtiEnv,
```

Comment:
Interesting read in the documentation:
```
Sent when a method causes the virtual machine to allocate an Object visible to Java programming
language code and the allocation is not detectable by other intrumentation mechanisms. 
Generally object allocation should be detected by instrumenting the bytecodes of allocating methods. 
Object allocation generated in native code by JNI function calls should be detected using JNI function interception. 
Some methods might not have associated bytecodes and are not native methods, they instead are executed directly by the VM. 
These methods should send this event. 
Virtual machines which are incapable of bytecode instrumentation for some or all of their methods can send this event.

Cases where this event would not be generated:
    Allocation due to bytecodes -- for example, the new and newarray VM instructions
```

So this is capturing only a small subset of all allocations.